sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -26,9 +26,188 @@
+            """
+            Takes the args/kwargs.values() and flattens them into a flat representation of (any) ShardedTensors and their indices.
+            """
+            t_i, t_vals = [], []
+            for i, arg in enumerate(items):
+                if isinstance(arg, ShardedTensor):
+                    t_i.append(i)
+                    t_vals.append(arg)
+                elif isinstance(arg, list) and all(
+                    isinstance(val, ShardedTensor) for val in arg
+                ):
+                    t_i.append([i] * len(arg))
+                    t_vals.extend(arg)
+            return t_i, t_vals
+
+        def unflatten_args(
+            items: Tuple | Dict, t_i: List[int | List[int]], t_vals: List[ShardedTensor]
+        ) -> Tuple[Tuple, Dict[str, Any]]:
+            """
+            Converts the flattened and (potentially) modified args or kwargs.values() back into the original structure.
+            """
+            i_lookup = (
+                list(range(len(items)))
+                if isinstance(items, tuple)
+                else list(items.keys())
+            )
+            new_items = list(items) if isinstance(items, tuple) else dict(items)
+
+            for i in t_i:
+                if isinstance(i, int):
+                    new_items[i_lookup[i]] = t_vals.pop(0)
+                else:  # List[int]
+                    _popped_vals = [t_vals.pop(0) for _ in range(len(i))]
+                    new_items[i_lookup[i[0]]] = items[i_lookup[i[0]]].__class__(
+                        _popped_vals
+                    )
+
+            if isinstance(new_items, list):
+                new_items = tuple(new_items)
+            return new_items
+
+        def func_wrapper(*args: Tuple, **kwargs: Dict[str, Any]):
+            """
+            Wraps each operation, f, in order to transfer all unpinned ShardedTensors in its input args/kwargs onto the same device,
+            then calls f with the modified args/kwargs, finally it will ensure the result is on the same devices as the input tensors.
+
+            If no ShardedTensors are present in the input, then no changes are made to input/output.
+            """
+            t_i_args, t_vals_args = flatten_args(args)
+            t_i_kwargs, t_vals_kwargs = flatten_args(list(kwargs.values()))
+            t_vals = t_vals_args + t_vals_kwargs
+
+            t_vals = transfer_if_needed(*t_vals)
+
+            args = unflatten_args(args, t_i_args, t_vals[: len(t_vals_args)])
+            kwargs = unflatten_args(kwargs, t_i_kwargs, t_vals[len(t_vals_args) :])
+            res = f(*args, **kwargs)
+            if isinstance(res, ShardedTensor) and len(t_vals) > 0:
+                pinned = (
+                    res.pinned
+                    or (len(t_vals) == 1 and t_vals[0].pinned)
+                    or "_like_" in f.__name__
+                )  # TODO: How to handle reshard_like (and future similar ones properly)
```

Comment:
@rsuderman

Reply:
Yeah we will probably have to custom handle reshard like. Its a pretty good example of a transfer based operation so not really feasible to rely on the wrapper.